### PR TITLE
[SPARK-7655][Core][SQL] Remove 'scala.concurrent.ExecutionContext.Implicits.global' in 'ask' and 'BroadcastHashJoin'

### PR DESCRIPTION
**Pull request by zsxwing**
Because both `AkkaRpcEndpointRef.ask` and `BroadcastHashJoin` uses `scala.concurrent.ExecutionContext.Implicits.global`. However, because the tasks in `BroadcastHashJoin` are usually long-running tasks, which will occupy all threads in `global`. Then `ask` cannot get a chance to process the replies.

For `ask`, actually the tasks are very simple, so we can use `MoreExecutors.sameThreadExecutor()`. For `BroadcastHashJoin`, it's better to use `ThreadUtils.newDaemonCachedThreadPool`.

### CODE REVIEW

**Comment by zsxwing** [comment]
@rxin @yhuai @marmbrus

**Comment by zsxwing** [comment]
cc @pwendell

**Comment by zsxwing** [comment]
I think this is the root cause of SPARK-7655. @yhuai, could you help verify it?

File: core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala [hunk 1]
```diff
@@ -247,6 +249,10 @@ private[spark] class AkkaRpcEnvFactory extends RpcEnvFactory {
   }
 }
 
+private[akka] object AkkaRpcEnv {
+  val askExecutionContext = ExecutionContext.fromExecutorService(MoreExecutors.sameThreadExecutor())
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala]: Does this mean that ask messages will just block the calling thread?

**Comment by pwendell** [comment]
I'm not an expert on this code but it LGTM - it seems like the goal here is to isolate the thread pool used by the broadcasts.

File: core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala [hunk 2]
```diff
@@ -305,7 +310,7 @@ private[akka] class AkkaRpcEndpointRef(
         }
       case AkkaFailure(e) =>
         Future.failed(e)
-    }.mapTo[T]
+    }(AkkaRpcEnv.askExecutionContext).mapTo[T]
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala]: @pwendell `AkkaRpcEnv.askExecutionContext` is only used for `Future.flatMap`. And the logic in `flatMap` never blocks. So it's safe to just run it in the caller's thread.

**Comment by zsxwing** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala [hunk 3]
```diff
@@ -74,3 +74,9 @@ case class BroadcastHashJoin(
     }
   }
 }
+
+object BroadcastHashJoin {
+
+  private val broadcastHashJoinExecutionContext = ExecutionContext.fromExecutorService(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: is this unlimited number of threads?

File: core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala (hunk 2 above)
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/akka/AkkaRpcEnv.scala]: i'd add some inline documentation saying this is running in the same thread.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala (hunk 3 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: Yes.

Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: @rxin I added two new methods to ThreadUtils and set the max thread number of `broadcastHashJoinExecutionContext` to 1024.

**Comment by rxin** [comment]
lgtm

**Pull request closed without merging**